File: library/src/main/java/com/benny/library/neobinding/bind/CommandBinding.java
package com.benny.library.neobinding.bind;

import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;

/**
 * Created by benny on 11/5/15.
 */
public class CommandBinding<T> {
    private String property;
    private Observable<?> trigger;
    private Action1<? super Boolean> canExecute = aBoolean -> {};
    private CommandHandler<Throwable, T> commandHandler;

    public CommandBinding(String property, Observable<?> trigger, CommandHandler<Throwable, T> commandHandler) {
        this.property = property;
        this.trigger = trigger;
        this.commandHandler = commandHandler;
    }

    public CommandBinding canExecute(Action1<? super Boolean> canExecute) {
        this.canExecute = canExecute;
        return this;
    }

    public String property() {
        return property;
    }

    public void bindTo(BindingContext<?> bindingContext, Command<T> command) {
        command.canExecute().compose(bindingContext.applyLifecycle())
                .doOnSubscribe(()->ViewModelBinder.LogBinding(property + " " + "can execute"))
                .doOnTerminate(()->ViewModelBinder.LogUnbinding(property + " " + "can execute"))
                .subscribe(canExecute);
        trigger.flatMap(v -> wrapError(command.execute())).compose(bindingContext.applyLifecycle())
                .doOnSubscribe(() -> ViewModelBinder.LogBinding(property + " " + "execute"))
                .doOnTerminate(() -> ViewModelBinder.LogUnbinding(property + " " + "execute"))
                .subscribe(commandHandler::onSuccess);
    }

    @SuppressWarnings("unchecked")
    Observable<T> wrapError(Observable execute) {
        return execute.onErrorResumeNext((Func1<Throwable, Observable>) throwable -> {
            commandHandler.onError(throwable);
            return Observable.empty();
        });
    }
}

File: library/src/main/kotlin/com/benny/library/neobinding/bind/CommandBinding.kt
package com.benny.library.neobinding.bind

import rx.Observable
import rx.functions.Action1

/**
 * Created by benny on 11/17/15.
 */

public class CommandBinding(key: String, val trigger: Observable<Unit>, val canExecute: Action1<in Boolean> = Action1 {}) : PropertyBinding() {
    public var key: String = key
    private set

    public fun prefix(prefix: String): CommandBinding {
        if(!prefix.isEmpty()) key = "$prefix.$key"
        return this
    }

    fun bindTo(bindingContext: BindingContext, command: Command) {
        trigger.compose(bindingContext.applyLifecycle<Unit>())
                .subscribe { command.execute(canExecute) }
    }
}
